aep/general/0162/aep.md.j2

```diff
@@ -1,5 +1,396 @@
+      }
+    }
+  }
+}
+```
+
+{% endtabs %}
+
+- The method **must** use the `POST` HTTP verb.
+- The method **should** return a resource revision.
+- The request message **must** have a `path` field, referring to the resource
+  revision whose configuration the resource should be rolled back to.
+  - The field **must** be [annotated as required][aip-203].
+  - The field **must** identify the [resource type][aip-123] that it
+    references.
+
+### Child resources
+
+Resources with a revision history **may** have child resources. If they do,
+there are two potential variants:
+
+- Child resources where each child resource is a child of the parent resource
+  as a whole.
+- Child resources where each child resource is a child of _a single revision
+  of_ the parent resource.
+
+APIs **should not** include multiple levels of resources with revisions, as
+this quickly becomes difficult to reason about.
+
+### Standard methods
+
+Any standard methods **must** implement the corresponding AIPs (AIP-131,
+AIP-132, AIP-133, AIP-134, AIP-135), with the following additional behaviors:
+
+- List methods: By default, revisions in the list response **must** be ordered
+  in reverse chronological order. User can supply `order_by` to override the
+  default behavior.
+- If the revision supports aliasing, a delete method with the resource name
+  of the alias (e.g. `revisions/1.0.2`) **must** remove the alias instead of
+  deleting the resource.
+
+As revisions are nested under the resource, also see [cascading delete][].
+
+## Rationale
+
+### Abstract revisions as nested collection
+
+Revisions being resources under nested collection make revisions a first class
+citizen.
+
+- Revisions can offer standard get, list, and delete methods.
+- It retains the flexibility of extending new fields to a revision in addition to
+  the resource message.
+
+### Output only resource configuration
+
+Although it was an option to have the revision take in the resource
+configuration as part of the create method, doing so would have allowed users to
+submit resource configuration for a revision that the resource was never in.
+
+`OUTPUT_ONLY` and requiring that a created revision represents the resource at
+current point in time eliminates that issue.
+
+## History
+
+### Switching from a collection extension to a subcollection
+
+In aip.dev prior to 2023-09, revisions were more like extension of an existing
+resource by using `@` symbol. List and delete revisions were custom methods on
+the resource collection. A single Get method was used to retrieve either the
+resource revision, or the resource.
+
+Its primary advantage was allowing a resource reference to seamlessly refer to
+a resource, or its revision.
+
+It also had several disadvantages:
+
+-  List revisions is a custom method (:listRevisions) on the resource collection
+-  Delete revision is a custom method on the resource collection
+-  Not visible in API discovery doc
+-  Resource ID cannot use `@`
```

Comment:
```suggestion
-  Resource IDs cannot use the `@` symbol.
```

FWIW, we worked around this at Roblox as follows:

* Only the last `@` is assumed to specify a rev ID.
* If you have a `@` in your resource ID and don't want a revision then you have to add `@latest` to the end.  (But this wouldn't work  if we had a different strategy for when to create revisions.)

Reply:
Sure, perhaps it's not the main reason. It's arguable that revisions are a carve-out and therefore able to not follow the AEPs in this regard - but you don't really want a lot of carve-outs in your spec.